#56040 site health persistent object cache

manuelRod opened the pull request:
Adding Persistent Object cache check to Site Health.

Trac ticket: [#56040](https://core.trac.wordpress.org/ticket/56040)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

manuelRod committed:
Adding persistent object cache test

manuelRod committed:
Merge branch 'trunk' into 56040-site-health-persistent-object-cache

manuelRod committed:
adding test

manuelRod committed:
removing performance-lab lang domain

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+	 */
+	public function get_test_health_persistent_object_cache() {
+		/**
+		 * Filter the action URL for the persistent object cache health check.
```
`Filters` (as per coding standards, third person singular verbs should be used in function descriptions)

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+	/**
+	 * Tests persistent_object_cache health check.
+	 *
+	 * @since 6.1
```
`@since 6.1.0` would be better

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+		/**
+		 * Filter the action URL for the persistent object cache health check.
+		 *
+		 * @since 6.1
```
`@since 6.1.0` would be better

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+		}
+
+		/**
+		 * Filter the second paragraph of the health check's description
```
"Filters" :)

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+		 *
+		 * Plugin authors may want to append notes (not replace) on why object caching is recommended for their plugin.
+		 *
+		 * @since 6.1
```
`@since 6.1.0` :)

audrasjb commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2848,4 +2953,128 @@ public function is_development_environment() {
+	/**
+	 * Determines whether to suggest using a persistent object cache.
+	 *
+	 * @since 6.1
```
`@since 6.1.0` :)

audrasjb reviewed (changes requested):
I reported a few nitpicks on the PR (I didn't mention all occurrences though) 😇

manuelRod committed:
nitpicks

manuelRod commented:
Thanks for the nitpicks @audrasjb, I've fixed them.

spacedmonkey reviewed (approved):
Looks good to me.

manuelRod commented:
I think this one is ready to be merged.
cc: @felixarntz

Clorith commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+		 * @param string $notes              The notes appended to the health check description.
+		 * @param array  $available_services The list of available persistent object cache services.
+		 */
+		$notes = apply_filters( 'site_status_persistent_object_cache_notes', $notes, $available_services );
```
This filter may not be needed, since all tests can have each individual field filtered either via PHP, or JavaScript, with the `site_status_test_result` filter (which is ran on each individual test).

Unless there's a specific need for a unique one for this test specifically? If that's the case, perhaps it may even be better to expand on the existing filter to also have one that includes the test name in it for more targeted filters, such as for example `site_status_{$test_name}_test_result` (in this case becoming `site_status_persistent_object_cache_test_result`) 🤔

manuelRod replied:
Hello, @Clorith Thanks for chiming in. Good point here. I'm gonna try to explain the reason why behind it. 
We are porting these modules from the [performance plugin](https://github.com/WordPress/performance/), on there, we decided to have as many filters as needed to allow customization to this level. As we were aiming for hosting especially to active use these filters to give targeted recommendations to their clients.

tillkruss replied:
@Clorith: This filter specifically was requested by @spacedmonkey or @felixarntz IIRC to allow hosting companies to easily add their own notes to this heath check about their object caching solution.

Clorith replied:
Makes sense in general, I'm just curious why only this test would need that special filter, is this something that should be generalized for all tests instead? And is there any reasoning for not using the existing filters (like what shortcomings were encountered, these may be shortcomings that exist for the other existing tests that we need to consider as well elsewhere).

tillkruss replied:
The `site_status_persistent_object_cache_notes` is specifically for hosting providers wanting to _only_ alter the health check `description` if the test is suggestion the use of an object cache. Using standard filters would make this quite verbose.

felixarntz replied:
+1 on keeping this filter to make it obvious and straightforward for e.g. hosting providers that would like to alter the notes. The general filter would technically work as you're outlining @Clorith, however I agree that this would be overly verbose. The notes in this Site Health check are explicitly intended to be filtered, similarly to how e.g. the PHP upgrade notice that was separately implemented a few years back allows its additional explanation to be filtered. When we eventually document the ability to filter the notes, it would be unnecessarily complex and confusing to have to explain how you have to look for the specific field to modify. In addition, by having this specific filter, we maintain the same security standard as for the default content, as the content will go through `wp_kses()` _after_ filtering. If developers used the general filter, they would have to take care of it yourself, essentially duplicating what core already does here - or worse (and maybe even more likely), they don't know and don't even use `wp_kses()`.

Clorith commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
+
+		$result['status']         = 'recommended';
+		$result['label']          = __( 'You should use a persistent object cache' );
+		$result['badge']['color'] = 'orange';
```
The badge should not have changing colors based on status, the color in core is used to indicate which area it belongs to. This is also a necessity for accessibility, where a color change alone is not sufficient, and the label would need to change with it if that were the case.

Instead, the `status` should change with the status of the test (which it does here, so this could be easily resolved by removing the badge color change).

manuelRod replied:
I didn't know that, so status is enough to get the right color. ok!

tillkruss replied:
@manuelRod: Can you update the PR, or do you want me to open on in the perflab repo first?

manuelRod replied:
@tillkruss should we do this first on the plugin level and then on the core level? or should we iterate directly here?

manuelRod replied:
@tillkruss @Clorith this is also being discussed in [here](https://github.com/WordPress/wordpress-develop/pull/2894#discussion_r941100596). We should get a decision that will affect both Site Health ( and future ) modules.

manuelRod replied:
fixed

felixarntz replied:
@manuelRod @tillkruss @Clorith I've opened https://github.com/WordPress/performance/issues/471 and https://github.com/WordPress/performance/pull/472 to fix this in the Performance Lab plugin throughout all Site Health checks.

Clorith commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2373,6 +2471,13 @@ public static function get_tests() {
 			);
 		}
 
+		if ( 'production' === wp_get_environment_type() ) {
```
Could we add an inline comment here, just to explain why the tests are only done in certain environments (I know it makes sense to us now, but it doesn't hurt to have the _why_ of it documented for the future).

tillkruss replied:
Makes sense.

>Only check for a persistent object cache in production environments to not unnecessarily promote complicated setups

Clorith commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2373,6 +2471,13 @@ public static function get_tests() {
 		}
 
+		if ( 'production' === wp_get_environment_type() ) {
+			$tests['direct']['persistent_object_cache'] = array(
```
Should this perhaps be an `async` test, given the potential DB load from `should_suggest_persistent_object_cache()`?

manuelRod replied:
bringing audrasjb here, since we wrote originally this module.

tillkruss replied:
@manuelRod: I wrote this 🙃

Async makes sense.

manuelRod replied:
ohh @tillkruss sorry for that, i mixed who wrote what :)

manuelRod replied:
@tillkruss should we do this kind of big change, first on the plugin level, and then move them here?

tillkruss replied:
Yeah, maybe we do the badge color and async change on `wordpress/performance` first? Do you want to open a PR, or shall I?

manuelRod replied:
@Clorith as @tillkruss comments in the next comment block, this query is super fast, I don't think we need to move to async either.

felixarntz replied:
Agreed, while I'm not opposed to making the check async, I don't think it's worth it either. Looking at existing Site Health checks, it appears like the policy for async tests is mostly around any external HTTP requests, which makes sense. Cheap database lookups like here however don't justify the additional effort IMO.

Clorith commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2848,4 +2953,128 @@ public function is_development_environment() {
+		// With InnoDB the `TABLE_ROWS` are estimates, which are accurate enough and faster to retrieve than individual `COUNT()` queries.
+		$results = $wpdb->get_results(
+			$wpdb->prepare(
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
```
I'd love for any line that disables coding standard checks to include a comment on _why_ the standards are being ignored, and why that is expected. I couldn't find any guidelines on this, and am checking in with the standards team to make sure we do this the right way :)

manuelRod replied:
I removed the second ignore which was not needed. However, this one is needed for 
`AND TABLE_NAME IN ('$table_names')`

@Clorith let me know when you hear from the standards team.

tillkruss replied:
@manuelRod: You removed the re-enable comment, which needs to stay.

Clorith replied:
There's no hard guideline on requiring omments per now, but the team agreed that this is something that should be created, so let's move forward with adding a comment on why a rule is being ignored here, be the change we want to see in the world 😃 

As for enabling/disabling of rules, you could also do a single-line-allow by using `// phpcs:ignore <rule name> -- <comment with reason for ignoring the rule right here>` for the single `$wpdb->prepare()` command, that may also be handy to not accidentally forget to re-enable a rule later on.

tillkruss replied:
@Clorith The wpdb query is extremely fast even on massive sites, I don't think we need to make this an async method.

tillkruss replied:
@manuelRod: Let's add the reason then. This query cannot use interpolation.

manuelRod replied:
Ignore comment added.

manuelRod replied:
fixed

Clorith reviewed (changes requested):
Had a look over, and I have some questions that I left in-line, and a quick reference for some potential improvements.

manuelRod committed:
nitpicks

manuelRod committed:
removing orange badge and relying on status only.
using phpcs:ignore rule for interpolation

manuelRod commented:
@tillkruss @Clorith I've implemented the changes.
1. Removed badge color for different statuses.
2. Added ignore comment for interpolation/CS.
3. Do we agree that the tests stay direct instead of async? the query is light and fast.
4. Do we agree on the extra filters to allow more flexibility to hosting companies?

Thanks!

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2373,6 +2470,14 @@ public static function get_tests() {
+		if ( 'production' === wp_get_environment_type() ) {
+			$tests['direct']['persistent_object_cache'] = array(
+				'label' => __( 'Persistent object cache' ),
+				'test'  => 'health_persistent_object_cache',
```
The `health_` prefix here and elsewhere does not make sense, let's remove it throughout this name and the relevant methods.
```suggestion
				'test'  => 'persistent_object_cache',
```

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
+	 *
+	 * @return array The health check result suggesting persistent object caching.
+	 */
+	public function get_test_health_persistent_object_cache() {
```
See my other comment:
```suggestion
	public function get_test_persistent_object_cache() {
```

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
 	}
 
+	/**
+	 * Tests persistent_object_cache health check.
```
This is not really an appropriate documentation :)

Please describe in one or two sentences what this really does, not just the name of the test it refers to, which is a bit pointless.

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return array The health check result suggesting persistent object caching.
```
It won't necessarily suggest persistent object caching - it might also say that everything is fine. I would suggest we rephrase it to the same description that is used in other core methods for Site Health tests.
```suggestion
	 * @return array The test result.
```

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
+		 */
+		$notes = apply_filters( 'site_status_persistent_object_cache_notes', $notes, $available_services );
+
+		$result['status']         = 'recommended';
+		$result['label']          = __( 'You should use a persistent object cache' );
+		$result['description']   .= sprintf(
```
I believe these are incorrectly indented? There are some additional spaces that should be removed.

felixarntz reviewed (commented):
@manuelRod Excellent work, this looks solid to me. I left a few comments on minor things to improve, such as the test slug and some documentation.

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2848,4 +2953,123 @@ public function is_development_environment() {
+	 *
+	 * @return array The list of available persistent object cache services.
+	 */
+	function available_object_cache_services() {
```
Methods should preferably always have an access modifier. Let's make this one `private`.

manuelRod committed:
nitpicks

manuelRod commented:
I think this is good to go now! cc: @Clorith @tillkruss @felixarntz

tillkruss commented:
LGTM

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,105 @@ public function get_test_authorization_header() {
+
+		$available_services = $this->available_object_cache_services();
+
+		$notes = __( 'Your hosting provider can tell you if persistent object caches are able to be enabled on your site.' );
```
This sounds a bit too complicated.
```suggestion
		$notes = __( 'Your hosting provider can tell you if persistent object caches can be enabled on your site.' );
```

tillkruss replied:
@felixarntz: Either `persistent object caching` or `a persistent object cache`.

manuelRod replied:
"Your hosting provider can tell you if persistent object cache can be enabled on your site."

felixarntz reviewed (approved):
@manuelRod One tiny nit-pick, however I can address that one pre-commit. This looks good to go, great work!

felixarntz commented:
@Clorith Can you please give this another review this week or early next week? It now has 3 approvals, from my perspective it's ready to commit.

manuelRod committed:
nitpick

tillkruss commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2253,6 +2253,105 @@ public function get_test_authorization_header() {
+
+		$available_services = $this->available_object_cache_services();
+
+		$notes = __( 'Your hosting provider can tell you if persistent object cache can be enabled on your site.' );
```
To me this sounds incorrect.

```suggestion
		$notes = __( 'Your hosting provider can tell you if a persistent object cache can be enabled on your site.' );
```

Clorith reviewed (approved):
I agree with the small string change by @tillkruss, but other than that it looks good 👍

felixarntz commented on src/wp-admin/includes/class-wp-site-health.php:
```diff
@@ -2848,4 +2955,123 @@ public function is_development_environment() {
+		 *
+		 * @param bool $suggest Whether to suggest using a persistent object cache.
+		 */
+		if ( apply_filters( 'site_status_suggest_persistent_object_cache', false ) ) {
```
@manuelRod A bit late, but I am just realizing that this should rather follow a more appropriate short-circuiting behavior: Right now, the only way to short-circuit would be to actually recommend a persistent object cache. But it also needs to be possible to run custom logic that would result in _not_ recommending a persistent object cache. So I think the default value here should be `null`, and any boolean result should cause the method to return early with that result.

This is a trivial change, which is compatible with the current behavior, it just expands it to also allow returning `false` to short-circuit.

felixarntz commented:
@manuelRod Committed in https://core.trac.wordpress.org/changeset/53955 🎉
